pkg/ai/customrest.go

```diff
@@ -0,0 +1,147 @@
+package ai
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+const CustomRestClientName = "customrest"
+
+type CustomRestClient struct {
+	nopCloser
+	client      *http.Client
+	base        *url.URL
+	token       string
+	model       string
+	temperature float32
+	topP        float32
+	topK        int32
+}
+
+type CustomRestRequest struct {
+	Model string `json:"model"`
+
+	// Prompt is the textual prompt to send to the model.
+	Prompt string `json:"prompt"`
+
+	// Options lists model-specific options. For example, temperature can be
+	// set through this field, if the model supports it.
+	Options map[string]interface{} `json:"options"`
+}
+
+type CustomRestResponse struct {
+	// Model is the model name that generated the response.
+	Model string `json:"model"`
+
+	// CreatedAt is the timestamp of the response.
+	CreatedAt time.Time `json:"created_at"`
+
+	// Response is the textual response itself.
+	Response string `json:"response"`
+}
+
+func (c *CustomRestClient) Configure(config IAIConfig) error {
+	baseURL := config.GetBaseURL()
+	if baseURL == "" {
+		baseURL = defaultBaseURL
+	}
+	c.token = config.GetPassword()
+	baseClientURL, err := url.Parse(baseURL)
+	if err != nil {
+		return err
+	}
+	c.base = baseClientURL
+
+	proxyEndpoint := config.GetProxyEndpoint()
+	c.client = http.DefaultClient
+	if proxyEndpoint != "" {
+		proxyUrl, err := url.Parse(proxyEndpoint)
+		if err != nil {
+			return err
+		}
+		transport := &http.Transport{
+			Proxy: http.ProxyURL(proxyUrl),
+		}
+
+		c.client = &http.Client{
+			Transport: transport,
+		}
+	}
+
+	c.model = config.GetModel()
+	if c.model == "" {
+		c.model = defaultModel
+	}
+	c.temperature = config.GetTemperature()
+	c.topP = config.GetTopP()
+	c.topK = config.GetTopK()
+	return nil
+}
+
+func (c *CustomRestClient) GetCompletion(ctx context.Context, prompt string) (string, error) {
+	var promptDetail struct {
+		Language string `json:"language,omitempty"`
+		Message  string `json:"message"`
+		Prompt   string `json:"prompt,omitempty"`
+	}
+	prompt = strings.ReplaceAll(prompt, "\n", "\\n")
+	prompt = strings.ReplaceAll(prompt, "\t", "\\t")
+	if err := json.Unmarshal([]byte(prompt), &promptDetail); err != nil {
+		return "", err
+	}
+	generateRequest := &CustomRestRequest{
+		Model:  c.model,
+		Prompt: promptDetail.Prompt,
+		Options: map[string]interface{}{
+			"temperature": c.temperature,
+			"top_p":       c.topP,
+			"top_k":       c.topK,
+			"message":     promptDetail.Message,
+			"language":    promptDetail.Language,
+		},
+	}
+	bts, err := json.Marshal(generateRequest)
```

Comment:
What does `bts` represent? Wouldn't it be better to use a more descriptive variable name?